core/include/vecmem/memory/atomic_ref.hpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+#if (defined(CL_SYCL_LANGUAGE_VERSION) || defined(SYCL_LANGUAGE_VERSION)) && \
+    defined(VECMEM_HAVE_SYCL_ATOMIC_REF)
+
+// SYCL include(s).
+#include <CL/sycl.hpp>
+
+namespace vecmem {
+
+/// @c vecmem::atomic_ref equals @c sycl::atomic_ref with "modern SYCL"
+template <typename T>
+using atomic_ref =
+    ::sycl::atomic_ref<T, ::sycl::memory_order::relaxed,
+                       ::sycl::memory_scope::device,
+                       ::sycl::access::address_space::global_space>;
+
+}  // namespace vecmem
+
+#elif ((!defined(__CUDA_ARCH__)) && (!defined(__HIP_DEVICE_COMPILE__)) && \
+       (!defined(CL_SYCL_LANGUAGE_VERSION)) &&                            \
+       (!defined(SYCL_LANGUAGE_VERSION)) && (__cplusplus >= 202000))
+
+// Syctem include(s).
+#include <atomic>
+
+namespace vecmem {
+
+/// @c vecmem::atomic_ref equals @c std::atomic_ref in host code with C++20
+template <typename T>
+using atomic_ref = std::atomic_ref<T>;
+
+}  // namespace vecmem
+
+#else
+
+// VecMem include(s).
+#include "vecmem/utils/types.hpp"
+
+// System include(s).
+#include <type_traits>
+
+namespace vecmem {
+
+/// Class providing atomic operations for the VecMem code
+///
+/// It is only meant to be used with primitive types. Ones that CUDA, HIP and
+/// SYCL built-in functions exist for. So no structs, or even pointers.
+///
+/// Note that it does not perform atomic operations in host code! That is only
+/// implemented with @c std::atomic_ref in C++20. With earlier C++ standards all
+/// operations in host code are performed as "regular" operations.
+///
+template <typename T>
+class atomic_ref {
+
+public:
+    /// @name Type definitions
+    /// @{
+
+    /// Type managed by the object
+    typedef T value_type;
+    /// Difference between two objects
+    typedef value_type difference_type;
+    /// Pointer to the value in global memory
+    typedef value_type* pointer;
+    /// Reference to a value given by the user
+    typedef value_type& reference;
+
+    /// @}
+
+    /// @name Check(s) on the value type
+    /// @{
+
+    static_assert(std::is_integral<value_type>::value,
+                  "vecmem::atomic_ref only accepts built-in integral types");
```

Comment:
In addition to the above, `std::atomic_ref` also supports floating point types, and in fact I think it supports any trivially copyable type. Is that not something we want to do in vecmem?

Reply:
Atomic float operations are only available above certain compute capabilities on NVIDIA hardware. But even more than that, not every single atomic operation is available for floating-point types even with the very latest hardware. See for instance:

https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicsub

In fact this class should probably be even more restrictive, saying that you should really only ever use it with `int` and `unsiged int`. Everything else may or may not work for your specific use case.

And no, I'm not interested in providing custom code in this project for emulating atomic operations on top of types that are not supported on the hardware level. Kokkos and LLAMA can have their fun with that...

Reply:
Furthermore, even "supported" implementations may be emulated via a CAS loop by hardware. For example, back in the Maxwell days, floating-point atomicAdd in shared memory was emulated, even though floating-point atomicAdd in global memory confusingly did have first-class hardware support (and good grief, because it is actually useful for many algorithms). Not sure if this was changed in newer hardware generations...